Mathlib/RingTheory/Ideal/AssociatedPrime/Finiteness.lean

```diff
@@ -0,0 +1,283 @@
+      |>.quotKerEquivOfSurjective ((LinearMap.range (auxLinearMap h))
+        |>.comapRestrict_surjective_of_surjective _ (auxSeq A M n).mkQ_surjective))
+  rw [Submodule.comapRestrict, LinearMap.ker_restrict, Submodule.ker_mkQ]
+  ext ⟨x, _⟩
+  simp only [LinearMap.mem_range, Subtype.exists, Submodule.mem_comap, Submodule.subtype_apply]
+  refine ⟨fun ⟨a, h1, h2⟩ ↦ ?_, fun h2 ↦ ⟨x, h2, rfl⟩⟩
+  rwa [← show a = x from congr($(h2).1)]
+
+end
+
+variable (A M)
+
+/-- If `A` is a Noetherian ring and `M` is a finitely generated `A`-module, then there exists
+a chain of submodules `0 = M₀ ≤ M₁ ≤ M₂ ≤ ... ≤ Mₙ = M` of `M`, such that for each `0 ≤ i < n`,
+`Mᵢ₊1 / Mᵢ` is isomorphic to `A / pᵢ` for some prime ideal `pᵢ` of `A`. -/
+@[stacks 00L0]
+theorem IsNoetherianRing.exists_relSeries_isQuotientEquivQuotientPrime :
+    ∃ s : RelSeries (Submodule.IsQuotientEquivQuotientPrime (A := A) (M := M)),
+      s.head = ⊥ ∧ s.last = ⊤ := by
+  have H : ∃ n : ℕ, auxSeq A M n = ⊤ := by
+    obtain ⟨n, h⟩ := monotone_stabilizes_iff_noetherian.2 (inferInstanceAs (IsNoetherian A M))
+      ⟨_, monotone_auxSeq A M⟩
+    exact ⟨n, auxSeq_eq_auxSeq_succ_iff.1 (h (n + 1) n.lt_add_one.le)⟩
+  classical exact ⟨⟨Nat.find H, fun ⟨n, _⟩ ↦ auxSeq A M n,
+    fun ⟨n, h⟩ ↦ isQuotientEquivQuotientPrime_auxSeq_of_ne_top (Nat.find_min H h)⟩,
+      auxSeq_zero A M, Nat.find_spec H⟩
+
+/-- If a property on finitely generated modules over a Noetherian ring satisfies that:
+
+- it holds for zero module (it's formalized as it holds for any module which is subsingleton),
+- it holds for `A ⧸ p` for every prime ideal `p` of `A` (to avoid universe problem,
+  it's formalized as it holds for any module isomorphic to `A ⧸ p`),
+- it is stable by short exact sequences,
+
+then the property holds for every finitely generated modules.
+
+NOTE: This should be the induction principle for `M`, but due to the bug
+https://github.com/leanprover/lean4/issues/4246
+currently it is induction for `Module.Finite A M`. -/
+@[elab_as_elim]
+theorem IsNoetherianRing.induction_on_isQuotientEquivQuotientPrime
+    ⦃M : Type v⦄ [AddCommGroup M] [Module A M] (_ : Module.Finite A M)
+    {motive : (N : Type v) → [AddCommGroup N] → [Module A N] → [Module.Finite A N] → Prop}
+    (subsingleton : (N : Type v) → [AddCommGroup N] → [Module A N] → [Module.Finite A N] →
+      [Subsingleton N] → motive N)
+    (quotient : (N : Type v) → [AddCommGroup N] → [Module A N] → [Module.Finite A N] →
+      (p : PrimeSpectrum A) → (N ≃ₗ[A] A ⧸ p.1) → motive N)
+    (exact : (N₁ : Type v) → [AddCommGroup N₁] → [Module A N₁] → [Module.Finite A N₁] →
+      (N₂ : Type v) → [AddCommGroup N₂] → [Module A N₂] → [Module.Finite A N₂] →
+      (N₃ : Type v) → [AddCommGroup N₃] → [Module A N₃] → [Module.Finite A N₃] →
+      (f : N₁ →ₗ[A] N₂) → (g : N₂ →ₗ[A] N₃) →
+      Function.Injective f → Function.Surjective g → Function.Exact f g →
+      motive N₁ → motive N₃ → motive N₂) : motive M := by
```

Comment:
Shouldn't `exact` imply that your `motive` is stable under isomorphisms and hence `quotient` should just be `motive (A / p)`?

Reply:
`motive (A / p)` can't compile since `A / p` is in different universe level. See discussions here [#mathlib4 > How to write induction principle for modules @ 💬](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/How.20to.20write.20induction.20principle.20for.20modules/near/513275096)

Reply:
This also explains why `exact` does not use `Submodule` and quotient.